pattern/constraint/value/AssignmentConstraint.java

```diff
@@ -0,0 +1,623 @@
+        Map<Identifier, Variable> argMap = new HashMap<>();
+        Set<Variable> argSet = new HashSet<>();
+        for (Variable v : constraint.variables()) {
+            Variable clonedVar = cloner.clone(v);
+            argMap.put(clonedVar.id(), clonedVar);
+            argSet.add(clonedVar);
+        }
+        AssignmentConstraint clone = new AssignmentConstraint(valueVariable, constraint.expression().clone(argMap), argSet);
+        clone.compiledExpression(constraint.compiledExpression());
+        return clone;
+    }
+
+    @Override
+    public Variable owner() {
+        return owner;
+    }
+
+    public Expression expression() {
+        return expression;
+    }
+
+    public Set<Variable> arguments() {
+        return arguments;
+    }
+
+    public Encoding.ValueType<?> valueType() {
+        return compiledExpression != null ? compiledExpression.returnType() : null;
+    }
+
+    public void compiledExpression(com.vaticle.typedb.core.traversal.expression.Expression<?> compiled) {
+        this.compiledExpression = compiled;
+    }
+
+    public com.vaticle.typedb.core.traversal.expression.Expression<?> compiledExpression() {
+        return compiledExpression;
+    }
+
+    @Override
+    public void addTo(GraphTraversal.Thing traversal) {
+        assert compiledExpression != null;
+        Set<Identifier.Variable> argIds = iterate(arguments).map(Variable::id).toSet();
+        traversal.assignment(owner.id(), argIds, compiledExpression);
+    }
+
+    @Override
+    public boolean isAssignment() {
+        return true;
+    }
+
+    @Override
+    public AssignmentConstraint asAssignment() {
+        return this;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+        AssignmentConstraint that = (AssignmentConstraint) o;
+        return (this.owner.equals(that.owner) && this.expression.equals(that.expression));
+    }
+
+    @Override
+    public int hashCode() {
+        return this.hash;
+    }
+
+    @Override
+    public Constraint clone(Conjunction.ConstraintCloner cloner) {
+        Set<Variable> clonedVariables = new HashSet<>();
+        arguments.forEach(v -> {
+            if (v.isThing()) clonedVariables.add(cloner.cloneVariable(v.asThing()));
+            else if (v.isType()) clonedVariables.add(cloner.cloneVariable(v.asType()));
+            else if (v.isValue()) clonedVariables.add(cloner.cloneVariable(v.asValue()));
+            else throw TypeDBException.of(ILLEGAL_STATE);
+        });
+        Map<Identifier, Variable> argMap = new HashMap<>();
+        clonedVariables.forEach(v -> argMap.put(v.id(), v));
+        AssignmentConstraint cloned = cloner.cloneVariable(this.owner).assign(expression.clone(argMap), clonedVariables);
+        cloned.compiledExpression(compiledExpression);
+        return cloned;
+    }
+
+    public abstract static class Expression {
```

Comment:
I don't think we gain anything out of having this instead of the TypeQL expression.

Reply:
I think you might be right - does it all work if we just keep the `compiledExpression` on the `AssignmentConstraint`?  What about the `toString`, should just print the TypeQL expression?

Reply:
I think the compiled 'traversal' level to-string can be printed in a very functional way just like we printed edges like
`--PLAYS-->` etc. So yes, the symbol should be there (eg. should be readable in a query plan).

The `toString` on `Constraint`s should be pretty much parse-able by TypeQL though. However, if we go with this then there is no Constraint-equivalent of Expressions. Thus, we would have to embed the TypeQL expressions & use their `toString()`. Would that work?

Reply:
We could indeed embed the TypeQL expression into the compiled expression, but we can't call toString, because it would recurse into the TypeQL sub-expressions, while we'd want it to recurse into compiled sub-expressions.

Do we really need traversal level expressions to be parseable by typeql?